### PR TITLE
Update left top rule order in constraint solving

### PR DESCRIPTION
**Pull request by leafpetersen**
Update to fix for https://github.com/dart-lang/language/issues/1035 .

### CODE REVIEW

**Comment by leafpetersen** [comment]
cc @scheglov @stefantsov

**Review by lrhn — APPROVED** [review]

**Comment by leafpetersen** [comment]
@stefantsov Good catch.   I worked through the rules, here's another take.  I've moved the left top rule into the right nullability rule.  The reasoning is that:
  - The right FutureOr rule might apply (this is your counter-example), in which case we want to match using that rule.  `dynamic` is a subtype match for `FutureOr<Q0>` iff `dynamic` is a subtype match for `Q0`.
  - The right nullability rule might apply, and this is where we actually need the left top rule, so I've added a clause there.  If `P` is `dynamic`, then the rule can match in one of two ways: either by matching `Object` against `Q0`, or by matching `dynamic` against `Q0`.  we prefer the former (heuristically), and fall back on the latter.
  
PTAL, thanks.

cc @scheglov

**Comment by scheglov** [comment]
I implemented the latest changes in the analyzer, but it affects inference for legacy libraries.
```dart
T f<T>(T a) => a;

test(dynamic a) {
  f(a);
}
```

We infer `Object*` for `T`.
```
[trySubtypeMatch][P: dynamic][Q: T*]
[trySubtypeMatch][P: dynamic][Q: T?]
[trySubtypeMatch][P: Object][Q: T]
[Object*]
```
I uploaded [work in progress CL](https://dart-review.googlesource.com/c/sdk/+/162621) to run it on bots, and see how many failures it has.

**Comment by stefantsov** [comment]
> I implemented the latest changes in the analyzer, but it affects inference for legacy libraries.
> 

FWIW, there are two separate implementations of the algorithm in the CFE — one for the legacy mode and one for the Null Safe mode — so the CFE isn't affected in the same way.

**Comment by stefantsov** [comment]
I filed an issue to update co19 tests according to the spec proposal: https://github.com/dart-lang/co19/issues/936.

**Comment by leafpetersen** [comment]
@scheglov That will almost certainly be true breaking.  Update with an additional clause to preserve the legacy behavior.  PTAL.

File: resources/type-system/inference.md [hunk 1]
```diff
@@ -774,7 +777,10 @@ clauses need be tried.
 - If `P` is a legacy type `P0*` then the match holds under constraint set `C`:
   - Only if `P0` is a subtype match for `Q` under constraint set `C`.
 - If `Q` is a legacy type `Q0*` then the match holds under constraint set `C`:
-  - Only if `P` is a subtype match for `Q?` under constraint set `C`.
+  - If `P` is `dynamic` or `void` and `P` is a subtype match for `Q` under
+    constraint set `C`.
```
Comment (stefantsov) [inline resources/type-system/inference.md:781]: It seems that this rule makes the algorithm recurse infinitely.

Comment (leafpetersen) [reply resources/type-system/inference.md:781]: Sorry, there was a typo in the original rule that I carried over to the new rule.  Fixed in both rules now.

**Comment by scheglov** [comment]
I landed the implementation into the analyzer, but we had to revert it because of breaking google3.

The case how it breaks is following.
Before:
```
[trySubtypeMatch][P: Actor* Function(Map<String*, Object*>*)*][Q: R* Function(Map<String*, dynamic>*)*]
[trySubtypeMatch][P: Actor* Function(Map<String*, Object*>*)][Q: R* Function(Map<String*, dynamic>*)*]
[trySubtypeMatch][P: Actor* Function(Map<String*, Object*>*)][Q: R* Function(Map<String*, dynamic>*)?]
[trySubtypeMatch][P: Actor* Function(Map<String*, Object*>*)][Q: R* Function(Map<String*, dynamic>*)]
[trySubtypeMatch][P: Actor*][Q: R*]
[trySubtypeMatch][P: Actor][Q: R*]
[trySubtypeMatch][P: Actor][Q: R?]
[trySubtypeMatch][P: Actor][Q: R]
[trySubtypeMatch][P: Map<String*, dynamic>*][Q: Map<String*, Object*>*]
[trySubtypeMatch][P: Map<String*, dynamic>][Q: Map<String*, Object*>*]
[trySubtypeMatch][P: Map<String*, dynamic>][Q: Map<String*, Object*>?]
[trySubtypeMatch][P: Map<String*, dynamic>][Q: Map<String*, Object*>]
[trySubtypeMatch][P: String*][Q: String*]
[trySubtypeMatch][P: dynamic][Q: Object*]
[trySubtypeMatch][P: dynamic][Q: Object?]
[trySubtypeMatch][P: dynamic][Q: Object]
[trySubtypeMatch][P: dynamic][Q: Null]
[trySubtypeMatch][P: Map<String*, dynamic>][Q: Null]
[inferrer][t1: Actor* Function(Map<String*, Object*>*)*][t2: R* Function(Map<String*, dynamic>*)*][success: true]
```

After:
```
[trySubtypeMatch][P: Actor* Function(Map<String*, Object*>*)*][Q: R* Function(Map<String*, dynamic>*)*]
[trySubtypeMatch][P: Actor* Function(Map<String*, Object*>*)][Q: R* Function(Map<String*, dynamic>*)*]
[trySubtypeMatch][P: Actor* Function(Map<String*, Object*>*)][Q: R* Function(Map<String*, dynamic>*)?]
[trySubtypeMatch][P: Actor* Function(Map<String*, Object*>*)][Q: R* Function(Map<String*, dynamic>*)]
[trySubtypeMatch][P: Actor*][Q: R*]
[trySubtypeMatch][P: Actor][Q: R*]
[trySubtypeMatch][P: Actor][Q: R?]
[trySubtypeMatch][P: Actor][Q: R]
[trySubtypeMatch][P: Map<String*, dynamic>*][Q: Map<String*, Object*>*]
[trySubtypeMatch][P: Map<String*, dynamic>][Q: Map<String*, Object*>*]
[trySubtypeMatch][P: Map<String*, dynamic>][Q: Map<String*, Object*>?]
[trySubtypeMatch][P: Map<String*, dynamic>][Q: Map<String*, Object*>]
[trySubtypeMatch][P: String*][Q: String*]
[trySubtypeMatch][P: dynamic][Q: Object*]
[trySubtypeMatch][P: dynamic][Q: Object]
[trySubtypeMatch][P: Map<String*, dynamic>][Q: Null]
[trySubtypeMatch][P: Actor* Function(Map<String*, Object*>*)][Q: Null]
[inferrer][t1: Actor* Function(Map<String*, Object*>*)*][t2: R* Function(Map<String*, dynamic>*)*][success: false]
```

So, we apply "If `P` is `dynamic` or `void` and `P` is a subtype match for `Q0` under constraint set `C`." and `dynamic` is not a subtype match for `Object`, so we fail.

**Comment by leafpetersen** [comment]
@scheglov I don't understand how the first trace is succeeding, can you elaborate?  Specifically, the diff between the two traces is that the original trace tries:

[trySubtypeMatch][P: dynamic][Q: Object*]
[trySubtypeMatch][P: dynamic][Q: Object?]
[trySubtypeMatch][P: dynamic][Q: Object]
[trySubtypeMatch][P: dynamic][Q: Null]

which looks like a failing trace to me: it tries to match `dynamic <# Object` and fails, then tries `dynamic <: Null` and fails.

The other trace looks like the following:

[trySubtypeMatch][P: String*][Q: String*]
[trySubtypeMatch][P: dynamic][Q: Object*]
[trySubtypeMatch][P: dynamic][Q: Object]

It's different, but gets the same answer: it just doesn't try to the `Null` branch that the other does.  This matches my intuition behind why I thought my fix was correct: asking `dynamic <: Q0?` can only succeed by matching `dynamic <: Q0`.

What am I missing?

**Comment by scheglov** [comment]
When we match `dynamic <# Object*` we now replace it with `dynamic <# Object` which fails.

But before this change we replace it with `dynamic <# Object?` which first tries `Q?` rules (and fails, while doing this fails `dynamic <# Object`, also then fails `dynamic <# Null`), but then we fall through from `Q?` and apply "If `Q` is `dynamic`, `Object?`, or `void` then the match holds under no no constraints", with succeeds.

**Comment by leafpetersen** [comment]
@scheglov I see, thanks.  I want to think this through again, but I'll follow up again shortly.

**Comment by leafpetersen** [comment]
OK, I've removed the restriction that I'd added on the right legacy rule, it was just there as an (incorrect) optimization.  PTAL.

**Comment by scheglov** [comment]
Thank you. The internal presubmit was green, so I landed [it](https://dart-review.googlesource.com/c/sdk/+/164240).

**Pull request merged**